platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/UnionFileCollection.java

```diff
@@ -81,4 +84,35 @@ protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
             visitor.accept(fileCollection);
         }
     }
+
+    @Override
+    public Optional<FileCollectionExecutionTimeValue> calculateExecutionTimeValue() {
+        Set<FileCollectionExecutionTimeValue> sourceExecutionTimeValues = new HashSet<>();
+        for (FileCollectionInternal fileCollection : source) {
+            Optional<FileCollectionExecutionTimeValue> executionTimeValue = fileCollection.calculateExecutionTimeValue();
+            if (!executionTimeValue.isPresent()) {
+                return Optional.empty();
+            }
+            sourceExecutionTimeValues.add(executionTimeValue.get());
+        }
+
+        return Optional.of(new UnionExecutionTimeValue(ImmutableSet.copyOf(sourceExecutionTimeValues)));
+    }
+
+    private static class UnionExecutionTimeValue implements FileCollectionExecutionTimeValue {
+        private final ImmutableSet<FileCollectionExecutionTimeValue> source;
+
+        public UnionExecutionTimeValue(ImmutableSet<FileCollectionExecutionTimeValue> source) {
+            this.source = source;
+        }
+
+        @Override
+        public FileCollectionInternal toFileCollection(FileCollectionFactory fileCollectionFactory) {
+            Set<FileCollectionInternal> source = this.source.stream()
+                .map(value -> value.toFileCollection(fileCollectionFactory))
+                .collect(Collectors.toSet());
```

Comment:
❓ Why do it in two steps?

Reply:
Did you mean something like
```
Set<FileCollectionInternal> source = this.source.stream()
                .collect(Collectors.mapping(
                    executionTimeValue -> executionTimeValue.toFileCollection(fileCollectionFactory),
                    Collectors.toSet()
                ))
```
?
If so, there is an actionable warning
<img width="611" alt="Screenshot 2024-07-22 at 17 22 02" src="https://github.com/user-attachments/assets/71fec6eb-a651-417e-898f-e3f9604329f2">